### PR TITLE
SLVS-1117: Rule S1451 - Track lack of copyright and license headers

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicense.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                    stac.ReportDiagnostic(Diagnostic.Create(Rule, Location.Create(stac.Tree, TextSpan.FromBounds(0, 0)), properties));
+                }
+            });
+        }
+
+        private static bool IsRegexPatternValid(string pattern)
+        {
+            try
+            {
+                new Regex(pattern);
```
Comment (SonarTech) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicense.cs]: CRITICAL Either remove this useless object instantiation of class "Regex" or use it. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1848)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicense.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        private static string GetHeaderOrDefault(SyntaxTrivia trivia)
+        {
+            var isComment = trivia.IsKind(SyntaxKind.SingleLineCommentTrivia) || trivia.IsKind(SyntaxKind.MultiLineCommentTrivia);
+
+            return isComment ? trivia.ToString() : null;
+        }
+
+        private static bool AreHeadersEqual(string header, string expectedHeader, bool isRegex)
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicense.cs]: It's strange to pass the `isRegex` here, when that's coming from an instance field. Let's make the method non-static, and then we can use the field directly.

The same is true for the `expectedHeader`.

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicense.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                return Regex.IsMatch(header, expectedHeader, RegexOptions.Compiled);
+            }
+            else
+            {
+                return header == expectedHeader;
+            }
+        }
+
+        private ImmutableDictionary<string, string> CreateProperties()
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicense.cs]: Let's rename this to `CreateDiagnosticProperties` to be more explicit. (Making it a property is also an option)

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        public sealed override FixAllProvider GetFixAllProvider() => DocumentBasedFixAllProvider.Instance;
+
+        protected sealed override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            if (!diagnostic.Properties.Any() ||
+                !diagnostic.Properties.ContainsKey(CheckFileLicense.IsRegularExpressionPropertyKey) ||
+                !diagnostic.Properties.ContainsKey(CheckFileLicense.HeaderFormatPropertyKey))
+            {
+                return Task.Delay(0);
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs]: We discussed it in person, but to me this is not easy to read. I would `return;` here and make the method async. You could also return `Task.CompletedTask` which shows the intent a bit better.

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                    Title,
+                    c =>
+                    {
+                        var fixedLeadingTrivias = FixNodeLeadingTrivias(syntaxNode.GetLeadingTrivia(), diagnostic.Properties[CheckFileLicense.HeaderFormatPropertyKey]);
+                        var newRoot = root.ReplaceNode(syntaxNode, syntaxNode.WithLeadingTrivia(fixedLeadingTrivias));
+                        return Task.FromResult(context.Document.WithSyntaxRoot(newRoot));
+                    }),
+                context.Diagnostics);
+
+            return Task.Delay(0);
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs]: if the method is async, then you can drop this line.

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        private IEnumerable<SyntaxTrivia> FixNodeLeadingTrivias(SyntaxTriviaList currentTriviaList, string comment)
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs]: this method can probably be `static`.

File: src/Tests/SonarAnalyzer.UnitTest/Common/RuleTest.cs [hunk 7]
```diff
@@ -28,6 +28,7 @@
 using SonarAnalyzer.Common;
 using Microsoft.CodeAnalysis.CodeFixes;
 using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Rules.CSharp;
```
Comment (tvsonar) [inline src/Tests/SonarAnalyzer.UnitTest/Common/RuleTest.cs]: This seems to be redundant.

File: src/SonarAnalyzer.ruleset [hunk 8]
```diff
@@ -17,6 +17,7 @@
     <Rule Id="S131" Action="Warning" />
     <Rule Id="S1313" Action="Warning" />
     <Rule Id="S134" Action="Warning" />
+    <Rule Id="S1451" Action="Warning" />
```
Comment (tvsonar) [inline src/SonarAnalyzer.ruleset]: This should also be added to the SonarAnalyzer.Test.ruleset.

File: src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+using System.Collections.Generic;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public class CheckFileLicenseCodeFixProvider : SonarCodeFixProvider
+    {
+        internal const string Title = "Add or update license header";
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CheckFileLicense.DiagnosticId);
+        public sealed override FixAllProvider GetFixAllProvider() => DocumentBasedFixAllProvider.Instance;
```
Comment (tvsonar) [inline src/SonarAnalyzer.CSharp/Rules/CheckFileLicenseCodeFixProvider.cs]: You can use the `WellKnownFixAllProvider` here, maybe that's faster.

**Pull request merged**